src/sage/crypto/public_key/key_exchange/key_exchange.py

```diff
@@ -0,0 +1,56 @@
+from abc import ABC, abstractmethod
+from sage.structure.sage_object import SageObject
+
+class KeyExchange(SageObject, ABC):
+
+    @abstractmethod
+    def alice_secret_key(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def alice_public_key(self, alice_secret_key):
+        raise NotImplementedError
+
+    @abstractmethod
+    def bob_secret_key(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def bob_public_key(self, bob_secret_key):
+        raise NotImplementedError
+
+    @abstractmethod
+    def alice_compute_shared_secret(self, alice_sk, bob_pk):
+        raise NotImplementedError
+
+    @abstractmethod
+    def bob_compute_shared_secret(self, bob_sk, alice_pk):
+        raise NotImplementedError
+
+    def alice_keygen(self):
+        alice_sk = self.alice_secret_key()
+        alice_pk = self.alice_public_key(alice_sk)
+        bob_sk = self.bob_secret_key()
+        bob_pk = self.bob_public_key(bob_sk)
+        shared_secret = self.alice_compute_shared_secret(alice_sk, bob_pk)
+        return (alice_sk, alice_pk, shared_secret)
+
+    def bob_keygen(self):
+        alice_sk = self.alice_secret_key()
+        alice_pk = self.alice_public_key(alice_sk)
+        bob_sk = self.bob_secret_key()
+        bob_pk = self.bob_public_key(bob_sk)
+        shared_secret = self.alice_compute_shared_secret(bob_sk, alice_pk)
+        return (bob_sk, bob_pk, shared_secret)
+
+    def compute_all_keys(self):
+        alice_sk = self.alice_secret_key()
+        alice_pk = self.alice_public_key(alice_sk)
+        bob_sk = self.bob_secret_key()
+        bob_pk = self.bob_public_key(bob_sk)
+        alice_shared_secret = self.alice_compute_shared_secret(alice_sk, bob_pk)
+        bob_shared_secret = self.bob_compute_shared_secret(bob_sk, alice_pk)
+        return (alice_sk, alice_pk, alice_shared_secret, bob_sk, bob_pk, bob_shared_secret)
+       
+
```

Comment:
```suggestion
    def _test_key_exchange(self, **options):
        tester = self._tester(**options)

        alice_sk, alice_pk = self.alice_keygen()                        
        bob_sk, bob_pk = self.bob_keygen()
        alice_shared_secret = self.alice_compute_shared_secret(alice_sk, bob_pk)
        bob_shared_secret = self.bob_compute_shared_secret(bob_sk, alice_pk)
        tester.assertEqual(alice_shared_secret, bob_shared_secret)
```
Run 5 key exchanges, error if any of them do not arrive at the same shared secret.

This will get picked up by `TestSuite`.

Reply:
This code looks good, I'll try and implement a wrong key exchange and see the error diagnostic but I believe that the `assertEqual` will only show the result of the `alice_shared_secret` and `bob_shared_secret` values, I'll look into the `assertEqual` docs, but I would like to also show what the public and secret keys were that lead to a bug.

Reply:
`assertEqual` has an optional `msg` keyword parameter, you can add the public and secret keys there

Reply:
https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual